File: raid_test.ts
//
// RAiD tests
//
import { assertEquals } from "@std/assert";

import {
  normalizeRAiD,
  normalizeRAiDShort,
  validateRAiD,
  validateRAiDShort,
} from "./raid.ts";
import { validateDOI } from "./doi.ts";

Deno.test("test normalizeRAiD", () => {
  const tests: Array<[string, string]> = [
    ["10.26259/0e59e9a5", "https://raid.org/10.26259/0e59e9a5"],
    ["102.26259/0e59e9a5", "https://raid.org/102.26259/0e59e9a5"],
    [
      "https://raid.org/10.26259/0e59e9a5",
      "https://raid.org/10.26259/0e59e9a5",
    ],
    [
      "  HTTPS://RAID.ORG/10.26259/0E59E9A5  ",
      "https://raid.org/10.26259/0e59e9a5",
    ],
  ];
  for (const [input, expected] of tests) {
    assertEquals(normalizeRAiD(input), expected);
  }
});

Deno.test("test normalizeRAiDShort", () => {
  const tests: Array<[string, string]> = [
    ["10.26259/0e59e9a5", "10.26259/0e59e9a5"],
    ["102.26259/0e59e9a5", "102.26259/0e59e9a5"],
    ["https://raid.org/10.26259/0e59e9a5", "10.26259/0e59e9a5"],
    ["https://raid.org/102.26259/0e59e9a5", "102.26259/0e59e9a5"],
    ["  HTTPS://RAID.ORG/10.26259/0E59E9A5  ", "10.26259/0e59e9a5"],
  ];
  for (const [input, expected] of tests) {
    assertEquals(normalizeRAiDShort(input), expected);
  }
});

Deno.test("test validateRAiD and validateRAiDShort", () => {
  const valid: string[] = [
    "10.26259/0e59e9a5",
    "102.26259/0e59e9a5",
    "https://raid.org/10.26259/0e59e9a5",
    "https://raid.org/102.26259/0e59e9a5",
    "HTTPS://RAID.ORG/10.26259/0E59E9A5",
  ];
  for (const id of valid) {
    assertEquals(validateRAiD(id), true, `expected validateRAiD(${id}) true`);
    assertEquals(
      validateRAiDShort(id),
      true,
      `expected validateRAiDShort(${id}) true`,
    );
  }

  const invalid: string[] = [
    "",
    "not-a-raid",
    "11.26259/0e59e9a5", // does not start with 10 or 102
  ];
  for (const id of invalid) {
    assertEquals(
      validateRAiD(id),
      false,
      `expected validateRAiD(${id}) false`,
    );
    assertEquals(
      validateRAiDShort(id),
      false,
      `expected validateRAiDShort(${id}) false`,
    );
  }
});

// Documents the relationship between DOI and RAiD identifiers: RAiD reuses
// DOI's "10.xxxx/yyyy" identifier space (RAiDs are issued as DOIs via
// DataCite), so a bare "10.xxxx/yyyy" is valid as both a DOI and a RAiD. The
// "102.xxxx/yyyy" form is RAiD-exclusive, since DOI requires a literal "10."
// prefix.
Deno.test("test DOI/RAiD disambiguation", () => {
  const shared = "10.26259/0e59e9a5";
  assertEquals(validateDOI(shared), true);
  assertEquals(validateRAiD(shared), true);

  const raidOnly = "102.26259/0e59e9a5";
  assertEquals(validateDOI(raidOnly), false);
  assertEquals(validateRAiD(raidOnly), true);
});

File: raid_test.go
package metadatatools

import (
	"testing"
)

func TestNormalizeRAiD(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"10.26259/0e59e9a5", "https://raid.org/10.26259/0e59e9a5"},
		{"10.83962/fb5be317", "https://raid.org/10.83962/fb5be317"},
		{"https://raid.org/10.26259/0e59e9a5", "https://raid.org/10.26259/0e59e9a5"},
		{"https://raid.org/10.83962/f2a7645d", "https://raid.org/10.83962/f2a7645d"},
		{"  HTTPS://RAID.ORG/10.26259/0E59E9A5  ", "https://raid.org/10.26259/0e59e9a5"},
	}
	for _, test := range tests {
		got := NormalizeRAiD(test.input)
		if got != test.expected {
			t.Errorf("NormalizeRAiD(%q) = %q, expected %q", test.input, got, test.expected)
		}
	}
}

func TestNormalizeRAiDShort(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"10.26259/0e59e9a5", "10.26259/0e59e9a5"},
		{"10.83962/fb5be317", "10.83962/fb5be317"},
		{"https://raid.org/10.26259/0e59e9a5", "10.26259/0e59e9a5"},
		{"https://raid.org/10.83962/f2a7645d", "10.83962/f2a7645d"},
		{"  HTTPS://RAID.ORG/10.26259/0E59E9A5  ", "10.26259/0e59e9a5"},
	}
	for _, test := range tests {
		got := NormalizeRAiDShort(test.input)
		if got != test.expected {
			t.Errorf("NormalizeRAiDShort(%q) = %q, expected %q", test.input, got, test.expected)
		}
	}
}

func TestValidateRAiD(t *testing.T) {
	valid := []string{
		"10.26259/0e59e9a5",
		"10.83962/fb5be317",
		"https://raid.org/10.26259/0e59e9a5",
		"https://raid.org/10.83962/f2a7645d",
		"HTTPS://RAID.ORG/10.26259/0E59E9A5",
	}
	for _, id := range valid {
		if !ValidateRAiD(id) {
			t.Errorf("ValidateRAiD(%q) = false, expected true", id)
		}
		if !ValidateRAiDShort(id) {
			t.Errorf("ValidateRAiDShort(%q) = false, expected true", id)
		}
	}

	invalid := []string{
		"",
		"not-a-raid",
		"11.26259/0e59e9a5",  // does not start with 10
		"102.26259/0e59e9a5", // "102." is not a real RAiD prefix, see D8
	}
	for _, id := range invalid {
		if ValidateRAiD(id) {
			t.Errorf("ValidateRAiD(%q) = true, expected false", id)
		}
		if ValidateRAiDShort(id) {
			t.Errorf("ValidateRAiDShort(%q) = true, expected false", id)
		}
	}
}

// TestRAiDIsDOIShaped documents the relationship between DOI and RAiD
// identifiers: RAiDs are issued as DOIs via DataCite, so RAiD and DOI share
// the same "10.<4-9 digit registrant code>/<suffix>" format. There is no
// format-level way to distinguish a RAiD from a DOI - any valid RAiD is also
// a valid DOI, and vice versa. See dev-notes/decisions_RAiD_support.md D8.
func TestRAiDIsDOIShaped(t *testing.T) {
	shared := []string{"10.26259/0e59e9a5", "10.83962/fb5be317"}
	for _, id := range shared {
		if !ValidateDOI(id) {
			t.Errorf("ValidateDOI(%q) = false, expected true (RAiD is DOI-shaped)", id)
		}
		if !ValidateRAiD(id) {
			t.Errorf("ValidateRAiD(%q) = false, expected true (RAiD is DOI-shaped)", id)
		}
	}
}
